### PR TITLE
ARROW-2122: [Python] Pyarrow fails to serialize dataframe with timestamp.

### PR DESCRIPTION
**Pull request by adshieh**
Fixes [ARROW-2122](https://issues.apache.org/jira/browse/ARROW-2122).

### CODE REVIEW

File: python/pyarrow/types.pxi [hunk 1]
```diff
@@ -847,6 +847,25 @@ cdef timeunit_to_string(TimeUnit unit):
         return 'ns'
 
 
+FIXED_OFFSET_PREFIX = '+'
```
Comment (adshieh) [inline python/pyarrow/types.pxi]: @wesm suggestions?

**Comment by wesm** [comment]
This needs a little more scrutiny (cc @jreback @cpcloud @pitrou) before we commit to something for 0.9.0 we might have to break later. I'm going to move this issue off 0.9.0 so we aren't blocking

File: python/pyarrow/types.pxi (hunk 1 above)
Comment (pitrou) [reply python/pyarrow/types.pxi]: We probably want the offset to be encoded as `[+-]HH:MM`.
See https://github.com/apache/arrow/blob/master/format/Schema.fbs#L162-L166

File: python/pyarrow/types.pxi [hunk 2]
```diff
@@ -847,6 +847,25 @@ cdef timeunit_to_string(TimeUnit unit):
...
+
+
+def tzinfo_to_string(tz):
```
Comment (pitrou) [inline python/pyarrow/types.pxi]: These two functions would deserve a docstring, and appropriate unit tests.

File: python/pyarrow/tests/test_convert_pandas.py [hunk 3]
```diff
@@ -1001,6 +1001,17 @@ def test_array_from_pandas_date_with_mask(self):
         assert pa.Array.from_pandas(expected).equals(result)
 
 
+def test_fixed_offset_timezone():
```
Comment (pitrou) [inline python/pyarrow/tests/test_convert_pandas.py]: Please put this under the class above.

File: python/pyarrow/types.pxi (hunk 1 above)
Comment (adshieh) [reply python/pyarrow/types.pxi]: Done!

File: python/pyarrow/types.pxi (hunk 2 above)
Comment (adshieh) [reply python/pyarrow/types.pxi]: Done!

File: python/pyarrow/tests/test_convert_pandas.py (hunk 3 above)
Comment (adshieh) [reply python/pyarrow/tests/test_convert_pandas.py]: Done!

**Review by pitrou — APPROVED** [review]
This looks good to me. Perhaps a pandas or timezone expert wants to comment, otherwise I think it's good to go.

File: python/pyarrow/types.pxi [hunk 4]
```diff
@@ -847,6 +849,63 @@ cdef timeunit_to_string(TimeUnit unit):
         return 'ns'
 
 
+_FIXED_OFFSET_RE = re.compile(r'([+-])(0[0-9]|1[0-9]|2[0-3]):([0-5][0-9])')
```
[the remark this answers is not in the document]
Comment (adshieh) [reply python/pyarrow/types.pxi]: Good catch, thanks!

**Review by xhochy — APPROVED** [review]
+1, LGTM

CI failures are unrelated.

**Pull request closed without merging**